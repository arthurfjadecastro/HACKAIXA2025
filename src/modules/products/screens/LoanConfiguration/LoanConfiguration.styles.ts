import { StyleSheet } from 'react-native';
import { colors } from '@/design-system/tokens';

export const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: colors.background.primary,
  },
  
  // Header
  header: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
    paddingHorizontal: 16,
    paddingVertical: 12,
    backgroundColor: colors.background.secondary,
    borderBottomWidth: 1,
    borderBottomColor: colors.border.light,
  },
  backButton: {
    padding: 16,
  },
  headerTitle: {
    fontSize: 14,
    fontWeight: '400',
    color: '#757575',
    textAlign: 'right',
    flex: 1,
    marginRight: 16,
  },
  headerSpacer: {
    width: 0,
  },

  // Content
  content: {
    flex: 1,
    paddingHorizontal: 16,
  },
  
  contentContainer: {
    paddingTop: 24,
    paddingBottom: 24,
  },

  // Amount Summary
  amountSummary: {
    marginBottom: 24,
  },
  amountRow: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
  },
  amountValue: {
    fontSize: 24,
    fontWeight: 'bold',
    color: colors.text.primary,
  },
  editButton: {
    padding: 8,
    backgroundColor: '#F5F5F5',
    borderRadius: 6,
  },

  // Installment Selector
  installmentSelector: {
    marginBottom: 24,
  },
  sectionLabel: {
    fontSize: 16,
    fontWeight: '600',
    color: colors.text.primary,
    marginBottom: 12,
  },
  controlsRow: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
    marginBottom: 8,
  },
  controlButton: {
    width: 40,
    height: 40,
    borderRadius: 20,
    backgroundColor: '#F5F5F5',
    justifyContent: 'center',
    alignItems: 'center',
    marginHorizontal: 16,
  },
  controlButtonDisabled: {
    backgroundColor: '#F0F0F0',
  },
  monthsInput: {
    width: 80,
    height: 40,
    borderWidth: 1,
    borderColor: '#E0E0E0',
    borderRadius: 8,
    textAlign: 'center',
    fontSize: 16,
    fontWeight: '600',
    color: colors.text.primary,
  },
  limitsHelper: {
    alignItems: 'center',
    marginTop: 8,
  },
  helperText: {
    fontSize: 12,
    color: '#BDBDBD',
  },

  // Installment Value
  installmentValue: {
    alignItems: 'center',
    marginBottom: 24,
  },
  installmentLabel: {
    fontSize: 16,
    color: colors.text.secondary,
    marginBottom: 4,
  },
  installmentAmount: {
    fontSize: 28,
    fontWeight: 'bold',
    color: '#1565C0',
  },

  // Financial Summary
  financialSummary: {
    backgroundColor: '#F8F9FA',
    borderRadius: 8,
    padding: 16,
    marginBottom: 24,
  },
  summaryRow: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    marginBottom: 8,
  },
  summaryLabel: {
    fontSize: 14,
    color: colors.text.secondary,
  },
  summaryValue: {
    fontSize: 14,
    fontWeight: '600',
    color: colors.text.primary,
  },

  // Footer
  footer: {
    paddingHorizontal: 16,
    paddingVertical: 16,
    backgroundColor: colors.background.secondary,
  },
  continueButton: {
    backgroundColor: '#F7931E',
    borderRadius: 8,
    height: 56,
    justifyContent: 'center',
  },
});
