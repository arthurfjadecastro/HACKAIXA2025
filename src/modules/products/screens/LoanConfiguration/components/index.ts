export { LoanHeader } from './LoanHeader';
export { AmountSummary } from './AmountSummary';
export { InstallmentSelector } from './InstallmentSelector';
export { FinancialSummary } from './FinancialSummary';
