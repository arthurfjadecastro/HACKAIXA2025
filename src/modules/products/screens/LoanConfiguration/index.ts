export { default } from './LoanConfiguration';
export * from './components';
export * from './hooks';
export * from './types';
