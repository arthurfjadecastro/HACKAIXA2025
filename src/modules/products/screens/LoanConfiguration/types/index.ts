export interface QuoteData {
  installment: number;
  total: number;
  rate: number;
}

export interface ProductData {
  maxMonths: number;
  minMonths: number;
  rateAm: number;
}

export interface ValidationState {
  showWarning: boolean;
  timeout: NodeJS.Timeout | null;
}

export interface LoanConfigurationProps {
  productId: string;
  amount: string;
}

export interface InstallmentSelectorProps {
  months: number;
  inputValue: string;
  productMinMonths: number;
  productMaxMonths: number;
  showValidationWarning: boolean;
  onMonthsChange: (value: string) => void;
  onDecrease: () => void;
  onIncrease: () => void;
}

export interface AmountSummaryProps {
  amount: string;
  onEdit: () => void;
}

export interface InstallmentValueProps {
  quote: QuoteData | null;
}

export interface FinancialSummaryProps {
  quote: QuoteData | null;
  productRateAm: number;
}

export interface LoanHeaderProps {
  onBackPress: () => void;
}
