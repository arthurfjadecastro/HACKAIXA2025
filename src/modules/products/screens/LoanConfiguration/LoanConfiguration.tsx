import React, { useState, useEffect } from 'react';
import { View, ScrollView } from 'react-native';
import { useRoute, RouteProp, useNavigation } from '@react-navigation/native';
import { NativeStackNavigationProp } from '@react-navigation/native-stack';
import { AppStackParamList } from '@/navigation/AppStack';

import { Text, Button } from '@/design-system/components';
import { useProducts } from '../../hooks';
import ProductDataService from '@/services/products/data';

// Importar componentes modulares
import { 
  LoanHeader, 
  AmountSummary, 
  InstallmentSelector, 
  FinancialSummary 
} from './components';
import { useLoanCalculation } from './hooks';
import { styles } from './LoanConfiguration.styles';

type RouteProps = RouteProp<AppStackParamList, 'LoanConfiguration'>;
type NavigationProps = NativeStackNavigationProp<AppStackParamList>;

const LoanConfiguration: React.FC = () => {
  const route = useRoute<RouteProps>();
  const navigation = useNavigation<NavigationProps>();
  const { productId, amount } = route.params;
  const { products } = useProducts();
  const productService = ProductDataService.getInstance();
  
  // Estado para dados do produto
  const [product, setProduct] = useState<any>(null);
  const [loading, setLoading] = useState(true);
  
  // Hook customizado para c√°lculos
  const {
    months,
    inputValue,
    quote,
    showValidationWarning,
    productMinMonths,
    productMaxMonths,
    productRateAm,
    formatCurrency,
    decreaseMonths,
    increaseMonths,
    handleMonthsChange,
    handleMonthsBlur,
  } = useLoanCalculation({ product, amount });

  // Carregar dados do produto
  useEffect(() => {
    const loadProductData = async () => {
      try {
        setLoading(true);
        console.log('üîç Carregando dados do produto:', productId);
        
        // Primeiro, busca o produto na lista de produtos cadastrados
        const registeredProduct = products.find(p => p.id === productId);
        
        if (registeredProduct) {
          // Busca os dados completos baseado no productId
          let detailedData = null;
          
          // Determina o tipo do produto baseado no ID ou nome (incluindo varia√ß√µes de acentua√ß√£o)
          const productNameLower = registeredProduct.name.toLowerCase();
          
          if (productId.includes('inss') || productNameLower.includes('inss')) {
            detailedData = await productService.loadConvenio('inss');
          } else if (productId.includes('militar') || productNameLower.includes('militar')) {
            detailedData = await productService.loadConvenio('militar');
          } else if (productId.includes('funcef') || productNameLower.includes('funcef')) {
            detailedData = await productService.loadConvenio('funcef');
          } else if (productId.includes('tjdft') || productNameLower.includes('tjdft')) {
            detailedData = await productService.loadConvenio('tjdft');
          } else if (productId.includes('habitacao') || 
                     productNameLower.includes('habitacao') || 
                     productNameLower.includes('habita√ß√£o')) {
            // Para habita√ß√£o, carrega os dados do SAC por padr√£o (ambos t√™m os mesmos limites de prazo)
            detailedData = await productService.loadHabitacao('sac');
          } else if (productId.includes('outro') || productNameLower.includes('outro') || 
                     registeredProduct.categoria === 'OUTRO') {
            // Para produtos OUTRO, usar dados espec√≠ficos do produto registrado
            setProduct(registeredProduct);
            return;
          }
          
          if (detailedData) {
            // Combinar dados do produto registrado com dados detalhados
            const productData = {
              ...registeredProduct,
              prazoMaximo: detailedData.prazo?.maximoMeses || registeredProduct.prazoMaximo,
              prazoMinimo: detailedData.prazo?.minimoMeses || registeredProduct.prazoMinimo,
              juros: detailedData.juros?.mensal || registeredProduct.juros,
            };
            setProduct(productData);
          } else {
            setProduct(registeredProduct);
          }
        }
      } catch (error) {
        console.error('‚ùå Erro ao carregar dados do produto:', error);
        setProduct(null);
      } finally {
        setLoading(false);
      }
    };

    if (products.length > 0) {
      loadProductData();
    }
  }, [productId, products]);

  // Navega√ß√£o
  const handleBackPress = () => {
    navigation.goBack();
  };

  const handleEditAmount = () => {
    navigation.navigate('CreateProduct');
  };

  const handleContinue = () => {
    if (!quote) return;
    
    navigation.navigate('SimulationLoading', {
      productId,
      amount,
      months
    });
  };

  if (loading) {
    return (
      <View style={styles.container}>
        <Text>Carregando...</Text>
      </View>
    );
  }

  if (!product) {
    return (
      <View style={styles.container}>
        <Text>Produto n√£o encontrado</Text>
      </View>
    );
  }

  return (
    <View style={styles.container}>
      {/* Header */}
      <LoanHeader 
        onGoBack={handleBackPress}
        productName="Quantos meses 2/2"
      />

      <ScrollView 
        style={styles.content} 
        contentContainerStyle={styles.contentContainer}
        showsVerticalScrollIndicator={false}
      >
        {/* Valor do empr√©stimo (resumo edit√°vel) */}
        <AmountSummary 
          amount={amount}
          onEditAmount={handleEditAmount}
        />

        {/* Seletor de parcelas */}
        <InstallmentSelector
          months={months}
          inputValue={inputValue}
          productMinMonths={productMinMonths}
          productMaxMonths={productMaxMonths}
          showValidationWarning={showValidationWarning}
          onDecrease={decreaseMonths}
          onIncrease={increaseMonths}
          onChangeText={handleMonthsChange}
          onBlur={handleMonthsBlur}
        />

        {/* Resumo financeiro */}
        <FinancialSummary
          quote={quote}
          productRateAm={productRateAm}
          formatCurrency={formatCurrency}
        />
      </ScrollView>

      {/* Footer com bot√£o de continuar */}
      <View style={styles.footer}>
        <Button
          title="Continuar"
          onPress={handleContinue}
          style={{
            ...styles.continueButton,
            backgroundColor: showValidationWarning ? '#D0D0D0' : '#F7931E'
          }}
          disabled={showValidationWarning || !quote}
        />
      </View>
    </View>
  );
};

export default LoanConfiguration;
