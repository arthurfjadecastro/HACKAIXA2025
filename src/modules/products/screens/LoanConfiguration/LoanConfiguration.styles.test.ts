import { styles } from './LoanConfiguration.styles';

describe('LoanConfiguration Styles', () => {
  it('should export styles object', () => {
    expect(styles).toBeDefined();
    expect(typeof styles).toBe('object');
  });

  it('should have container style', () => {
    expect(styles.container).toBeDefined();
  });

  it('should have content style', () => {
    expect(styles.content).toBeDefined();
  });

  it('should have header style', () => {
    expect(styles.header).toBeDefined();
  });

  it('should match snapshot', () => {
    expect(styles).toMatchSnapshot();
  });
});
