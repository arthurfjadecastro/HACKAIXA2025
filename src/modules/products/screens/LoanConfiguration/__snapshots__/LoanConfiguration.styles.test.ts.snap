// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoanConfiguration Styles should match snapshot 1`] = `
{
  "amountRow": {
    "alignItems": "center",
    "flexDirection": "row",
    "justifyContent": "space-between",
  },
  "amountSummary": {
    "marginBottom": 24,
  },
  "amountValue": {
    "color": "#12161C",
    "fontSize": 24,
    "fontWeight": "bold",
  },
  "backButton": {
    "padding": 16,
  },
  "container": {
    "backgroundColor": "#F4F8FB",
    "flex": 1,
  },
  "content": {
    "flex": 1,
    "paddingHorizontal": 16,
  },
  "contentContainer": {
    "paddingBottom": 24,
    "paddingTop": 24,
  },
  "continueButton": {
    "backgroundColor": "#F7931E",
    "borderRadius": 8,
    "height": 56,
    "justifyContent": "center",
  },
  "controlButton": {
    "alignItems": "center",
    "backgroundColor": "#F5F5F5",
    "borderRadius": 20,
    "height": 40,
    "justifyContent": "center",
    "marginHorizontal": 16,
    "width": 40,
  },
  "controlButtonDisabled": {
    "backgroundColor": "#F0F0F0",
  },
  "controlsRow": {
    "alignItems": "center",
    "flexDirection": "row",
    "justifyContent": "center",
    "marginBottom": 8,
  },
  "editButton": {
    "backgroundColor": "#F5F5F5",
    "borderRadius": 6,
    "padding": 8,
  },
  "financialSummary": {
    "backgroundColor": "#F8F9FA",
    "borderRadius": 8,
    "marginBottom": 24,
    "padding": 16,
  },
  "footer": {
    "backgroundColor": "#FFFFFF",
    "paddingHorizontal": 16,
    "paddingVertical": 16,
  },
  "header": {
    "alignItems": "center",
    "backgroundColor": "#FFFFFF",
    "borderBottomColor": "#EEEEEE",
    "borderBottomWidth": 1,
    "flexDirection": "row",
    "justifyContent": "space-between",
    "paddingHorizontal": 16,
    "paddingVertical": 12,
  },
  "headerSpacer": {
    "width": 0,
  },
  "headerTitle": {
    "color": "#757575",
    "flex": 1,
    "fontSize": 14,
    "fontWeight": "400",
    "marginRight": 16,
    "textAlign": "right",
  },
  "helperText": {
    "color": "#BDBDBD",
    "fontSize": 12,
  },
  "installmentAmount": {
    "color": "#1565C0",
    "fontSize": 28,
    "fontWeight": "bold",
  },
  "installmentLabel": {
    "color": "#666666",
    "fontSize": 16,
    "marginBottom": 4,
  },
  "installmentSelector": {
    "marginBottom": 24,
  },
  "installmentValue": {
    "alignItems": "center",
    "marginBottom": 24,
  },
  "limitsHelper": {
    "alignItems": "center",
    "marginTop": 8,
  },
  "monthsInput": {
    "borderColor": "#E0E0E0",
    "borderRadius": 8,
    "borderWidth": 1,
    "color": "#12161C",
    "fontSize": 16,
    "fontWeight": "600",
    "height": 40,
    "textAlign": "center",
    "width": 80,
  },
  "sectionLabel": {
    "color": "#12161C",
    "fontSize": 16,
    "fontWeight": "600",
    "marginBottom": 12,
  },
  "summaryLabel": {
    "color": "#666666",
    "fontSize": 14,
  },
  "summaryRow": {
    "alignItems": "center",
    "flexDirection": "row",
    "justifyContent": "space-between",
    "marginBottom": 8,
  },
  "summaryValue": {
    "color": "#12161C",
    "fontSize": 14,
    "fontWeight": "600",
  },
}
`;
